### ISSUE TITLE
UseSnakeCaseNamingConvention() Does not Handle Owned Entities Table Names

### ISSUE THREAD

**Issue by twinmind** [efcore/EFCore.NamingConventions#33]
For example, for a context like this

```csharp
    public class Account
    {
        public Guid Id { get; set; }
        public string Email { get; set; }
        public byte[] PasswordHash { get; set; }
        public List<RefreshToken> RefreshTokens { get; set; }

        public bool OwnsToken(string token) 
        {
            return this.RefreshTokens?.Find(x => x.Token == token) != null;
        }
    }

    [Owned]
    public class RefreshToken
    {
        [Key]
        public Guid Id { get; set; }
        public Account Account { get; set; }
        public string Token { get; set; }
        public DateTime Expires { get; set; }
    }

    public class DataContext : DbContext
    {
        public DbSet<Account> Accounts { get; set; }

        public DataContext(DbContextOptions<DataContext> options) : base(options)
        {

        }

        protected override void OnConfiguring(DbContextOptionsBuilder options)
        {
            options.UseSnakeCaseNamingConvention();
        }
    }
```

I get initial migration with table names like this:

```csharp
        protected override void Up(MigrationBuilder migrationBuilder)
        {
            migrationBuilder.CreateTable(
                name: "accounts",
                columns: table => new
                {
                    id = table.Column<Guid>(nullable: false),      
                    email = table.Column<string>(nullable: true),
                    password_hash = table.Column<byte[]>(nullable: true)
                },
                constraints: table =>
                {
                    table.PrimaryKey("pk_accounts", x => x.id);
                });

            migrationBuilder.CreateTable(
                name: "RefreshToken",
                columns: table => new
                {
                    id = table.Column<Guid>(nullable: false),
                    account_id = table.Column<Guid>(nullable: false),
                    token = table.Column<string>(nullable: true),
                    expires = table.Column<DateTime>(nullable: false)
                },
                constraints: table =>
                {
                    table.PrimaryKey("pk_refresh_token", x => x.id);
                    table.ForeignKey(
                        name: "fk_refresh_token_accounts_account_id",
                        column: x => x.account_id,
                        principalTable: "accounts",
                        principalColumn: "id",
                        onDelete: ReferentialAction.Cascade);
                });

            migrationBuilder.CreateIndex(
                name: "ix_refresh_token_account_id",
                table: "RefreshToken",
                column: "account_id");
        }
```
I would expect table name to be `refresh_tokens` for entity `RefreshToken`. And it would be generated correctly if I remove `[Owned]` attribute on `RefreshToken` entity.

**Comment by roji** [comment]
Which versions of EF Core and of EFCore.NamingConventions are you using? Can you please try with version 5.0.0-preview5 of this library if you aren't using that?

**Comment by twinmind** [comment]
I'm using EFCore.NamingConventions 1.1.0 with Microsoft.EntityFrameworkCore 3.1.5 and Microsoft.EntityFrameworkCore 3.1.4 when I have this issue.
I tried to upgrade EFCore.NamingConventions to version 5.0.0-preview5 in my small ASP.NET Core project, but it seems that I need to upgrade other dependencies to 5.0.0 previews as well and also install actual .NET 5 SDK. I'll do that will update this issue.

**Comment by twinmind** [comment]
I've created a test console app, where I created my context, all references are 5.0.0 latest previews. When I run `add-migration initial`, I get runtime exception:
```
System.MissingMethodException: Method not found: 'System.String Microsoft.EntityFrameworkCore.RelationalEntityTypeExtensions.GetDefaultTableName(Microsoft.EntityFrameworkCore.Metadata.IEntityType)'.
   at EFCore.NamingConventions.Internal.NameRewriterBase.ProcessForeignKeyOwnershipChanged(IConventionForeignKeyBuilder relationshipBuilder, IConventionContext`1 context)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ImmediateConventionScope.OnForeignKeyOwnershipChanged(IConventionForeignKeyBuilder relationshipBuilder)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.OnForeignKeyOwnershipChangedNode.Run(ConventionDispatcher dispatcher)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.DelayedConventionScope.Run(ConventionDispatcher dispatcher)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ConventionBatch.Run()
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ConventionBatch.Dispose()
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.ImmediateConventionScope.OnModelInitialized(IConventionModelBuilder modelBuilder)
   at Microsoft.EntityFrameworkCore.Metadata.Conventions.Internal.ConventionDispatcher.OnModelInitialized(IConventionModelBuilder modelBuilder)
   at Microsoft.EntityFrameworkCore.Metadata.Internal.Model..ctor(ConventionSet conventions)
   at Microsoft.EntityFrameworkCore.ModelBuilder..ctor(ConventionSet conventions)
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.CreateModel(DbContext context, IConventionSetBuilder conventionSetBuilder)
   at Microsoft.EntityFrameworkCore.Infrastructure.ModelSource.GetModel(DbContext context, IConventionSetBuilder conventionSetBuilder)
   at Microsoft.EntityFrameworkCore.Internal.DbContextServices.CreateModel()
   at Microsoft.EntityFrameworkCore.Internal.DbContextServices.get_Model()
   at Microsoft.EntityFrameworkCore.Infrastructure.EntityFrameworkServicesBuilder.<>c.<TryAddCoreServices>b__7_3(IServiceProvider p)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactory(FactoryCallSite factoryCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSiteMain(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitCache(ServiceCallSite callSite, RuntimeResolverContext context, ServiceProviderEngineScope serviceProviderEngine, RuntimeResolverLock lockType)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScopeCache(ServiceCallSite singletonCallSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.DynamicServiceProviderEngine.<>c__DisplayClass1_0.<RealizeService>b__0(ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngine.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.ServiceProviderEngineScope.GetService(Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Microsoft.EntityFrameworkCore.DbContext.get_DbContextDependencies()
   at Microsoft.EntityFrameworkCore.DbContext.get_InternalServiceProvider()
   at Microsoft.EntityFrameworkCore.DbContext.Microsoft.EntityFrameworkCore.Infrastructure.IInfrastructure<System.IServiceProvider>.get_Instance()
   at Microsoft.EntityFrameworkCore.Infrastructure.Internal.InfrastructureExtensions.GetService[TService](IInfrastructure`1 accessor)
   at Microsoft.EntityFrameworkCore.Infrastructure.AccessorExtensions.GetService[TService](IInfrastructure`1 accessor)
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.CreateContext(Func`1 factory)
   at Microsoft.EntityFrameworkCore.Design.Internal.DbContextOperations.CreateContext(String contextType)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.RemoveMigration(String contextType, Boolean force)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.RemoveMigrationImpl(String contextType, Boolean force)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.RemoveMigration.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.<>c__DisplayClass3_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
Method not found: 'System.String Microsoft.EntityFrameworkCore.RelationalEntityTypeExtensions.GetDefaultTableName(Microsoft.EntityFrameworkCore.Metadata.IEntityType)'.
```

References:
![image](https://user-images.githubusercontent.com/9801734/87229432-b63bc780-c35c-11ea-8414-b5dee270a529.png)


If I remove `.UseSnakeCaseNamingConvention()`, then migration succeeds.

**Comment by twinmind** [comment]
With that being said, if this is reproducible in EFCore.NamingConventions 1.1.0, is there a possibility to release a minor version with the fix? I guess most people prefer to stick to stable releases.

**Comment by roji** [comment]
I'm guessing you're mixing preview versions, i.e. preview5 of this lib with preview6 of EF Core. Can you please try aligning to preview5?

**Comment by twinmind** [comment]
Good catch, thanks. I aligned all references to preview 5 and was able to generate migration. The same issue occurs with version 5.0.0-preview5 as with 1.1.0, i.e. owned entity table name doesn't get transformed according to convention. One other issue that came up is that non-key columns in owned entity table are prefixed with table name in snake case, i.e. `refresh_token_token`, `refresh_token_expires`, columns `id` and `account_id` are PK and FK respectively and don't have such prefix. Attached a screenshot with migration for the db context and entities from my original post.
![image](https://user-images.githubusercontent.com/9801734/87235714-b60aee80-c393-11ea-87ca-19509f677d74.png)

**Comment by roji** [comment]
Thanks @twinmind, I'll try take a look soon. There have been many changes from 3.1 to 5.0, so unfortunately I'm not sure I'll have time to release a fixed version that works with 3.1 - but let's see.

**Comment by lajith111** [comment]
i am using ef core 5  preview 7 .. migration  not working for owned type.. am getting below error..Same works previous version..
Please update to new version ..thanks

![image](https://user-images.githubusercontent.com/28708173/89725027-6d913000-da28-11ea-8c16-8e2d934bc4b3.png)

**Comment by roji** [comment]
I've just published 5.0.0-preview8 to work with the latest EF Core preview. As far as I know owned entities should work well in that version, so I'm going to close this issue for now. If you still encounter issues with this, please open a new issue with the full repro information for 5.0.0-preview8 and I'll handle it.

**Comment by lajith111** [comment]
No issue still persist in 5.0.0-rc.  After i run Update-Database -Context CBSContext
If i remove  options.UseSnakeCaseNamingConvention(); migration works fine ...i think same key name generration..Please check

```
Npgsql.PostgresException (0x80004005): 42P07: relation "pk_accounts" already exists
   at Npgsql.NpgsqlConnector.<>c__DisplayClass160_0.<<DoReadMessage>g__ReadMessageLong|0>d.MoveNext()
--- End of stack trace from previous location ---
   at Npgsql.NpgsqlConnector.<>c__DisplayClass160_0.<<DoReadMessage>g__ReadMessageLong|0>d.MoveNext()
--- End of stack trace from previous location ---
   at Npgsql.NpgsqlDataReader.NextResult(Boolean async, Boolean isConsuming)
   at Npgsql.NpgsqlDataReader.NextResult()
   at Npgsql.NpgsqlCommand.ExecuteReaderAsync(CommandBehavior behavior, Boolean async, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteNonQuery(Boolean async, CancellationToken cancellationToken)
   at Npgsql.NpgsqlCommand.ExecuteNonQuery()
   at Microsoft.EntityFrameworkCore.Storage.RelationalCommand.ExecuteNonQuery(RelationalCommandParameterObject parameterObject)
   at Microsoft.EntityFrameworkCore.Migrations.MigrationCommand.ExecuteNonQuery(IRelationalConnection connection, IReadOnlyDictionary`2 parameterValues)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.MigrationCommandExecutor.ExecuteNonQuery(IEnumerable`1 migrationCommands, IRelationalConnection connection)
   at Microsoft.EntityFrameworkCore.Migrations.Internal.Migrator.Migrate(String targetMigration)
   at Microsoft.EntityFrameworkCore.Design.Internal.MigrationsOperations.UpdateDatabase(String targetMigration, String connectionString, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabaseImpl(String targetMigration, String connectionString, String contextType)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.UpdateDatabase.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
  Exception data:
    Severity: ERROR
    SqlState: 42P07
    MessageText: relation "pk_accounts" already exists
    File: d:\pginstaller_13.auto\postgres.windows-x64\src\backend\catalog\index.c
    Line: 866
    Routine: index_create
42P07: relation "pk_accounts" already exists
```

**Comment by andrecasteliano** [comment]
The issue still exists on release 5.0.2
if you put [Owned] on a type, the naming convention does not get applied to the table.

**Comment by roji** [comment]
@andrecasteliano and others, I've indeed located an issue when an owned entity is used in a *collection* navigation (e.g. OwnsMany). If that's your case, then this is tracked by #69 and fixed in 6.0.0. Otherwise, please open a new issue with a full repro code so I can investigate.
